SB-26355 - Assessment score migration job Implementation

manjudr opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-26355" title="SB-26355" target="_blank"><img alt="Story" src="https://project-sunbird.atlassian.net/secure/viewavatar?size=medium&avatarId=10315&avatarType=issuetype" />SB-24793</a>  Score Migration Script
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
Assessment Aggregator Flink Job Enhancement:

• Enable an instruction event to recompute the best score and trigger certificate preprocessor event.

Cleanup Invalid Assessment Data:

1. Load the assessment agg data to dataframe for selected batches (from config). Use QUORUM read and write. - DataFrame1
2. Delete Invalid attempts - DataFrame2
• Create a dataframe from above(step 1 - DataFrame1), where the total max score is not the same as the assessment totalScore value (records with invalid total max score.)
• Use this dataframe and delete the records.
3. Recompute the best score
a. Generate an instruction event using DataFrame2 to recompute the score and issue a certificate if valid.
b. Write it to a file to push it manually to Kafka.
4. Revoke certificate data for enrolments which don't match the criteria.
a. Use DataFrame2 and join(left) with the user_enrolments table.
b. explode the issued_certificates column to extract the certificate identifier.
c. Generate a CSV  (userid, courseid, batchid, certificateid) to delete below data manually (or using another script).

Output of the above script will be two files.

1. A file with enrolment details and existing certificate id. Use these details to delete the below data.
a. user_enrolments - issued_certificates column to null
b. certificates ES documents
2. A file containing event to push it to Kafka (for assessment aggregator job)
a. push this to kafka using console-producer after deleting the above data (step 1).
### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

manjudr committed:
Issue #000 feat: Assessment Score Correction Job

manjudr committed:
Issue SB-24793 feat:  Assessment score migration Job

manjudr committed:
Issue SB-24793 feat:  Assessment score migration Job

manjudr committed:
Issue SB-24793 feat:  Enabling the code coverage

manjudr committed:
Issue #DIK-5641 feat:  Review comments chagnes - Correcting the score, max_score, and grand_total value

manjudr committed:
Issue #DIK-5641 feat:  Reverting unwanted changes

manjudr committed:
Issue #DIK-5641 feat:  Enabling the code coverage

manjudr committed:
Issue #DIK-5641 feat:  Save csv file issue fix

manjudr committed:
Issue SB-24793 feat:  Assessment score migration data product job implementation

manjudr committed:
Issue SB-24793 feat:  Assessment score correction cleanup

manjudr committed:
Issue SB-24793 feat:  Assessment score correction cleanup

manjudr committed:
Issue SB-24793 feat:  Score metrics correction script update

manjudr committed:
Issue SB-24793 feat:  Score metrics correction script update

manjudr committed:
Issue SB-24793 feat:  Score metrics correction script job config update

SanthoshVasabhaktula commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,109 @@
+    // Get the Assessment Data for the specific Batch
+    val assessmentData: DataFrame = getAssessmentAggData(batchId).select("course_id", "batch_id", "content_id", "attempt_id", "user_id", "total_max_score", "total_score").persist()
+    // Take the content Id which is associated to the batch being invoked for the correction
+    val contentId: String = assessmentData.select("content_id").collect().map(_ (0)).toList.head.asInstanceOf[String]
+
+    val contentMetaURL: String = modelParams.getOrElse("contentReadAPI", "https://diksha.gov.in/api/content/v1/read/").asInstanceOf[String]
```
This logic assumes that the course has only assessment and that too it is self assess?

manjudr replied:
Yes santhosh

SanthoshVasabhaktula replied:
There can be multiple assessments. Update the logic to get the first self-assess content

SanthoshVasabhaktula commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,109 @@
+      JobLogger.log("Generated the CSV File", Option(Map("batch_id" -> batchId, "total_records" -> totalRecords), INFO))
+    } else {
+      JobLogger.log("Deleting the records from the table", Option(Map("total_records" -> totalRecords)), INFO)
+      filteredAssessmentData.select("course_id", "batch_id", "user_id", "content_id", "attempt_id").rdd.deleteFromCassandra(AppConf.getConfig("sunbird.courses.keyspace"), "assessment_aggregator")
```
Where is the logic to revert certificates and recompute aggregates?

maheshkumargangula replied:
1. We need to get the confirmation from Krishna to revert certificates.
2. Updating the aggregates to user_activity_agg table missed here. We will add that logic.

SanthoshVasabhaktula replied:
This is not resolved yet?

manjudr replied:
Point 2. Implemented the logic - (Updating the aggregates to user_activity_agg table missed here. We will add that logic.)

manjudr committed:
Issue SB-24793 feat:  Score migration testcase issue fix

manjudr committed:
Issue SB-24793 feat:  Score migration testcase issue fix

manjudr committed:
Issue SB-24793 feat:  Code review comments changes -
			1. Updating the user activity aggregate  table agg column and agg_last_updated column for the score correction after removing the data from assessment agg table			2. Correcting only SelfAssess ContentType contents only.

manjudr committed:
Issue SB-24793 feat:  Code review comments changes -
			1. Updating the user activity aggregate  table agg column and agg_last_updated column for the score correction after removing the data from assessment agg table			2. Correcting only SelfAssess ContentType contents only.

manjudr committed:
Issue SB-24793 feat: Certificate deletion logic implementation

manjudr committed:
Issue SB-24793 feat: Certificate deletion logic implementation

manjudr committed:
Issue SB-24793 feat: Certificate deletion logic implementation

manjudr committed:
Issue SB-24793 feat: assessment record deletion

manjudr committed:
Issue SB-24793 feat: Assessment logic change for removing invalid attempt and generating the instruction events to correct the certificate

manjudr committed:
Issue SB-24793 feat: Assessment logic change for removing invalid attempt and generating the instruction events to correct the certificate

manjudr committed:
Issue SB-24793 feat: Adding write consistency level to the job.

manjudr committed:
Issue SB-24793 feat: Testcase fixes

manjudr committed:
Issue SB-24793 feat: Code coverage update

maheshkumargangula commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,187 @@
+    val incorrectRecords: DataFrame = assessmentDF.filter(col("content_id") === contentId).filter(col("total_max_score") =!= totalQuestions)
+      .select("course_id", "batch_id", "content_id", "attempt_id", "user_id", "total_max_score", "total_score")
+    val incorrectRecordsSize: Long = incorrectRecords.count()
+    JobLogger.log("Total Incorrect Records", Option(Map("total_records" -> incorrectRecordsSize)), INFO)
```
Include the complete context in logging. batchId, contentId and totalQuestions.

maheshkumargangula commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,187 @@
+                folderName: String)(implicit config: JobConfig): Unit = {
+    JobLogger.log("Generating the CSV File", None, INFO)
+    val outputPath = config.modelParams.getOrElse(Map[String, Option[AnyRef]]()).getOrElse("csvPath", "").asInstanceOf[String]
+    data.repartition(1).write.option("header", value = true).format("com.databricks.spark.csv").save(outputPath.concat(s"/$folderName-$batchId-${System.currentTimeMillis()}.csv"))
```
Please use the contentId also in the name of the file. If a course has multiple SelfAssess content, it will be difficult to understand for which content the file generated.

manjudr committed:
Issue SB-24793 feat: Code review comments fixes

manjudr committed:
Issue SB-24793 feat: removed the unwanted files

manjudr committed:
Issue SB-24793 feat: Added the error handling in the file write

maheshkumargangula commented:
@manjudr - the PR looks fine now.

anandp504 commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,200 @@
+                  totalQuestions: Int,
+                  isDryRunMode: Boolean
+                 )(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig, sc: SparkContext): AssessmentCorrectionMetrics = {
+    val incorrectRecords: DataFrame = assessmentDF.filter(col("content_id") === contentId).filter(col("total_max_score") =!= totalQuestions)
```
This logic of comparing total_max_score with totalQuestions is incorrect unless each question has a score of 1. Please add a test case to test this incorrect comparison.

anandp504 commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,200 @@
+                                contentId: String)(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig, sc: SparkContext): Unit = {
+    val outputPath = config.modelParams.getOrElse(Map[String, Option[AnyRef]]()).getOrElse("csvPath", "").asInstanceOf[String]
+    val file = new File(outputPath.concat(s"/instruction-events-$batchId-$contentId-${System.currentTimeMillis()}.json"))
+    var writer: BufferedWriter = null
```
The BufferedWriter can be created outside of the try-catch-finally block. Also, each line need not be flushed to the file since we are already using a BufferedWriter.

anandp504 commented on data-products/src/main/scala/org/sunbird/analytics/audit/AssessmentScoreCorrectionJob.scala:
```diff
@@ -0,0 +1,200 @@
+  }
+
+  // Method to fetch the totalQuestion count from the content meta
+  def getTotalQuestions(contentId: String,
```
Why is the getTotalQuestions encompassed inside a Future? We are making a blocking call to the getTotalQuestions inside the process function and hence a Future is not required.

manjudr committed:
Issue SB-24793 feat: Code review comments fixes -
1. Improved the Testcase assertions
2. Removed the Future In the API call logic
3. Filtering the invalid assessment based the count of the question
4. Updated the write flush logic.

anandp504 reviewed (approved):
Approved :+1:
